airflow/cli/commands/dag_command.py

```diff
@@ -497,6 +523,60 @@ def dag_test(args, session=None):
             print(dot_graph.source)
 
 
+@provide_session
+def _run_task(ti: TaskInstance, session=None):
+    """
+    Run a single task instance, and push result to Xcom for downstream tasks. Bypasses a lot of
+    extra steps used in `task.run` to keep our local running as fast as possible
+    Args:
+        ti: TaskInstance to run
+    """
+    current_task = ti.render_templates(ti.get_template_context())
+    log.info("*****************************************************")
+    log.info("Running task %s", current_task.task_id)
+    try:
+        xcom_value = current_task.execute(context=ti.get_template_context())
```

Comment:
```suggestion
        xcom_value = ti.task.execute(context=ti.get_template_context(session=session))
```

Reply:
You also aren't calling pre_execute or post_execute here, so some operators might not work.

You probaly want to call a method on TI instead.

Possible `ti._run_raw_task(session=session, test_mode=True)`

Reply:
@ashb I agree on using the _run_raw_task function, but ironically the test_mode might not be the best idea here. test_mode prevents us from recording passing/failure in the database, and without that state we won't know when to run downstream tasks.

Reply:
That's not _quite_ what test mode does.

It simply doesn't a `session.commit()`. But if you put a `session.flush()` after it (and are careful about transactions/passing session everywhere) then you can run _all_ tasks in a single transaction, never commit it, and roll it back at the end.

I just tested this change and it works for downstream. See the video below.

Reply:
@ashb when I run this with test mode it seems to infinite loop as the tasks are never marked as finished?